### PR TITLE
Fixes #36239 - Template load_resource - explain :joins, :preload and includes

### PR DESCRIPTION
**Pull request by stejskalleos**
**How to test**
* Go to `/templates_doc/v1/reports/Foreman::Renderer::Scope::Macros::Loaders/load_hosts.en.html`

**Screenshot**
eeeeeeeeeee

**Sauce:**
* https://www.bigbinary.com/blog/preload-vs-eager-load-vs-joins-vs-includes
* https://tadhao.medium.com/joins-vs-preload-vs-includes-vs-eager-load-in-rails-5f721c44b3a9

### CODE REVIEW

File: app/services/foreman/renderer/scope/macros/loaders.rb [hunk 1]
```diff
@@ -45,6 +45,25 @@ module Loaders
...
+# `where` cannot be used for associated data filtering
+Host.preload(:interfaces)
+# => SELECT \"hosts\".* FROM \"hosts\" WHERE \"hosts\".\"type\" = 'Host::Managed'
+# => SELECT \"nics\".* FROM \"nics\" WHERE \"nics\".\"host_id\" IN (...) ORDER BY \"nics\".\"identifier\" ASC
+
+# Includes - for loading, comparing and filtering data with associated data
+# Loads associated data with either a single query for simple cases (in terms of joining table)
+# or with two separate queries for other cases
+# `where` can be used for associated data filtering
+# Doesn't have the (n+1) issue
```
Comment (ianballou) [inline app/services/foreman/renderer/scope/macros/loaders.rb]: What is this about n+1 issue ?

**Comment by ofedoren** [comment]
> **How to test**
> 
> * `FOREMAN_APIPIE_LANGS=en bundle exec rake apipie:cache`

I don't think this will do anything, since we disabled cache for apipie docs: https://github.com/theforeman/foreman/blob/develop/config/initializers/apipie.rb#L43

File: app/services/foreman/renderer/scope/macros/loaders.rb (hunk 1 above)
Comment (stejskalleos) [reply app/services/foreman/renderer/scope/macros/loaders.rb]: I forgot to mention it also for `join` and `preload`. I guess we don't have to mention it.

**Comment by stejskalleos** [comment]
Rebased & updated
@ianballou @ofedoren any other comments?

**Review by ekohl — COMMENTED** [review]
Wouldn't it be better to refer to the official docs on this? https://guides.rubyonrails.org/active_record_querying.html#eager-loading-associations already documents this. It's a leaky abstraction, but then again the whole API is exactly that.

**Review by ofedoren — DISMISSED** [review]
I'm kinda OK with the descriptions. Maybe we could add something like "For more information, please refer official Rails documentation: `link`".

**Comment by stejskalleos** [comment]
Updated, I reduced the information so it won't take so  much space on the page and added link to official docs

**Review by ofedoren — DISMISSED** [review]
I'm okay with this, unless @ekohl has something to add, I'd merge.

File: app/services/foreman/renderer/scope/macros/loaders.rb [hunk 2]
```diff
@@ -45,6 +45,17 @@ module Loaders
               example "<% #{name}.each_record do |#{model.to_s.downcase}| %>
   <%= #{model.to_s.downcase}.id %>, <%= #{model.to_s.downcase}.#{model == User ? 'login' : 'name'} %>
 <% end %>", desc: "Prints #{model.to_s.downcase} id and #{model == User ? 'login' : 'name'} of each #{model.to_s.downcase}"
+              example "# Joins - for comparing and filtering data
```
Comment (ekohl) [inline app/services/foreman/renderer/scope/macros/loaders.rb]: Would it be good to use heredoc?
```suggestion
              example <<~EXAMPLE
                # Joins - for comparing and filtering data
```

Or at least single quotes so you don't have to escape `"` inline.

**Review by ekohl — COMMENTED** [review]
Textual :+1: but some minor coding suggestion.

**Review by ofedoren — APPROVED** [review]
Thanks, @stejskalleos and the crew :)

**Pull request merged**